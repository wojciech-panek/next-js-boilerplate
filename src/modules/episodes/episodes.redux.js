import { createActions, createReducer } from 'reduxsauce';
import { Record, List, fromJS, Map } from 'immutable';


export const { Types: EpisodesTypes, Creators: EpisodesActions } = createActions({
  fetch: [],
  fetchSuccess: ['data'],
  fetchError: ['payload'],
  fetchSingle: ['id'],
  fetchSingleSuccess: ['data'],
  fetchSingleError: ['payload'],
}, { prefix: 'EPISODES_' });

const EpisodesRecord = new Record({
  items: List(),
  data: Map(),
});

export const INITIAL_STATE = new EpisodesRecord({});

const getSuccessHandler = (state = INITIAL_STATE, action) => state.set('items', fromJS(action.data));

const getSingleSuccessHandler = (state = INITIAL_STATE, action) => state.set('data', fromJS(action.data));

export const reducer = createReducer(INITIAL_STATE, {
  [EpisodesTypes.FETCH_SUCCESS]: getSuccessHandler,
  [EpisodesTypes.FETCH_SINGLE_SUCCESS]: getSingleSuccessHandler,
});
