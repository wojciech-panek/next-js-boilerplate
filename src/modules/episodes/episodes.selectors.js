import { createSelector } from 'reselect';


const selectEpisodesDomain = state => state.get('episodes');

export const selectEpisodesItems = createSelector(
  selectEpisodesDomain, state => state.get('items')
);

export const selectEpisodesData = createSelector(
  selectEpisodesDomain, state => state.get('data')
);
