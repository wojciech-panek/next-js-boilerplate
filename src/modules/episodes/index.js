export {
  EpisodesTypes,
  EpisodesActions,
} from './episodes.redux';

export * from './episodes.selectors';
