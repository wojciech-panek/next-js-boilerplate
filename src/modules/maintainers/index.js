export {
  MaintainersTypes,
  MaintainersActions,
} from './maintainers.redux';

export * from './maintainers.selectors';
