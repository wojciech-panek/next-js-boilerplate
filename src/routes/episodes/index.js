export { default as Episodes } from './episodes.container';
