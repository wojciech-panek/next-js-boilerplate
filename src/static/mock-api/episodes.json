[
  {
    "id": 1,
    "series": 1,
    "number": "01",
    "title": "DII1 Episode 1 Title"
  },
  {
    "id": 2,
    "series": 1,
    "number": "02",
    "title": "DII1 Episode 2 Title"
  },
  {
    "id": 3,
    "series": 1,
    "number": "03",
    "title": "DII1 Episode 3 Title"
  },
  {
    "id": 4,
    "series": 2,
    "number": "01",
    "title": "DII2 Episode 1 Title"
  },
  {
    "id": 5,
    "series": 2,
    "number": "02",
    "title": "DII2 Episode 2 Title"
  },
  {
    "id": 6,
    "series": 2,
    "number": "03",
    "title": "DII2 Episode 3 Title"
  },
  {
    "id": 7,
    "series": 3,
    "number": "01",
    "title": "Liars Episode 1 Title"
  },
  {
    "id": 8,
    "series": 3,
    "number": "02",
    "title": "Liars Episode 2 Title"
  },
  {
    "id": 9,
    "series": 3,
    "number": "03",
    "title": "Liars Episode 3 Title"
  },
  {
    "id": 10,
    "series": 4,
    "number": "01",
    "title": "FML Episode 1 Title"
  },
  {
    "id": 11,
    "series": 4,
    "number": "02",
    "title": "FML Episode 2 Title"
  },
  {
    "id": 12,
    "series": 4,
    "number": "03",
    "title": "FML Episode 3 Title"
  }
]
