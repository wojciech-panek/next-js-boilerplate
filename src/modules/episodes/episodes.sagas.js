import { put, takeLatest } from 'redux-saga/effects';
import { delay } from 'redux-saga';

import reportError from '../../shared/utils/reportError';
import api from '../../shared/services/api';
import { EpisodesTypes, EpisodesActions } from './episodes.redux';


export function* fetchEpisodes() {
  try {
    yield delay(1000);
    const { data } = yield api.get('/static/mock-api/episodes.json');

    return yield put(EpisodesActions.fetchSuccess(data));
  } catch (e) {
    if (e.response) {
      return yield put(EpisodesActions.fetchError(e.response.data));
    }

    return yield reportError(e);
  }
}

export function* fetchSingleEpisode({ id }) {
  try {
    yield delay(1000);
    const { data: episodesData } = yield api.get('/static/mock-api/episodes.json');

    const data = episodesData.find((series) => series.id === id);

    return yield put(EpisodesActions.fetchSingleSuccess(data));
  } catch (e) {
    if (e.response) {
      return yield put(EpisodesActions.fetchSingleError(e.response.data));
    }

    return yield reportError(e);
  }
}

export function* watchEpisodes() {
  try {
    yield takeLatest(EpisodesTypes.FETCH, fetchEpisodes);
    yield takeLatest(EpisodesTypes.FETCH_SINGLE, fetchSingleEpisode);
  } catch (e) {
    yield reportError(e);
  }
}
