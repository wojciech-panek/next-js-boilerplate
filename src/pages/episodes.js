import React, { PureComponent } from 'react';
import { compose } from 'ramda';
import { withRouter } from 'next/router';

import withIntl from '../shared/lib/withIntl';
import { Episodes } from '../routes/episodes';
import { EpisodesActions } from '../modules/episodes';


export class EpisodesPage extends PureComponent {
  static fetchData(dispatch, query) {
    dispatch(EpisodesActions.fetchSingle(parseInt(query.id, 10)));
  }

  static async getInitialProps({ isServer, store, query }) {
    if (isServer) {
      await store.execSagaTasks(isServer, (dispatch) => {
        EpisodesPage.fetchData(dispatch, query);
      });
    } else {
      EpisodesPage.fetchData(store.dispatch, query);
    }

    return {};
  }

  render() {
    return (
      <Episodes />
    );
  }
}

export default compose(
  withIntl,
  withRouter,
)(EpisodesPage);
