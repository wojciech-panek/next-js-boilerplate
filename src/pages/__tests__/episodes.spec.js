import React from 'react';
import { shallow } from 'enzyme';
import { spy } from 'sinon';
import { expect } from 'chai';

import { EpisodesPage } from '../episodes';
import { EpisodesActions } from '../../modules/episodes';


describe('EpisodesPage: Page', () => {
  const defaultProps = {
  };

  const component = (props) => (
    <EpisodesPage {...defaultProps} {...props} />
  );

  it('should render correctly', () => {
    const wrapper = shallow(component());
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch EpisodesActions.fetchSingle in getInitialProps on server', async () => {
    const dispatch = spy();
    const query = { id: '1' };
    const store = global.utils.prepareInitialPropsStore(dispatch);

    await EpisodesPage.getInitialProps({ isServer: true, store, query });

    expect(dispatch).to.have.been.calledOnce;
    expect(dispatch).to.have.been.calledWith(EpisodesActions.fetchSingle(1));
  });

  it('should dispatch EpisodesActions.fetchSingle in getInitialProps on client', async () => {
    const query = { id: '1' };
    const store = {
      dispatch: spy(),
    };

    await EpisodesPage.getInitialProps({ isServer: false, store, query });

    expect(store.dispatch).to.have.been.calledOnce;
    expect(store.dispatch).to.have.been.calledWith(EpisodesActions.fetchSingle(1));
  });
});
