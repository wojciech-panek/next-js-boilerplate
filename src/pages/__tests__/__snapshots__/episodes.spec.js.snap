// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodesPage: Page should render correctly 1`] = `<InjectIntl(withRouter(Connect(Episodes))) />`;
